htlcswitch/hop/error_encryptor.go

```diff
@@ -66,35 +89,126 @@ type ErrorEncrypter interface {
 	// given io.Reader.
 	Decode(io.Reader) error
 
-	// Reextract rederives the encrypter using the extracter, performing an
-	// ECDH with the sphinx router's key and the ephemeral public key.
+	// Reextract rederives the encrypter using the shared secret generator,
+	// performing an ECDH with the sphinx router's key and the ephemeral
+	// public key.
 	//
 	// NOTE: This should be called shortly after Decode to properly
 	// reinitialize the error encrypter.
-	Reextract(ErrorEncrypterExtracter) error
+	Reextract(SharedSecretGenerator) error
 }
 
 // SphinxErrorEncrypter is a concrete implementation of both the ErrorEncrypter
 // interface backed by an implementation of the Sphinx packet format. As a
 // result, all errors handled are themselves wrapped in layers of onion
 // encryption and must be treated as such accordingly.
 type SphinxErrorEncrypter struct {
-	*sphinx.OnionErrorEncrypter
+	encrypter interface{}
 
 	EphemeralKey *btcec.PublicKey
+	CreatedAt    time.Time
+	errorStruct  *sphinx.AttributableErrorStructure
 }
 
-// NewSphinxErrorEncrypter initializes a blank sphinx error encrypter, that
-// should be used to deserialize an encoded SphinxErrorEncrypter. Since the
-// actual encrypter is not stored in plaintext while at rest, reconstructing the
-// error encrypter requires:
+// NewSphinxErrorEncrypterUninitialized initializes a blank sphinx error
+// encrypter, that should be used to deserialize an encoded
+// SphinxErrorEncrypter. Since the actual encrypter is not stored in plaintext
+// while at rest, reconstructing the error encrypter requires:
 //  1. Decode: to deserialize the ephemeral public key.
 //  2. Reextract: to "unlock" the actual error encrypter using an active
 //     OnionProcessor.
-func NewSphinxErrorEncrypter() *SphinxErrorEncrypter {
+func NewSphinxErrorEncrypterUninitialized() *SphinxErrorEncrypter {
 	return &SphinxErrorEncrypter{
-		OnionErrorEncrypter: nil,
-		EphemeralKey:        &btcec.PublicKey{},
+		EphemeralKey: &btcec.PublicKey{},
+	}
+}
+
+func NewSphinxErrorEncrypter(ephemeralKey *btcec.PublicKey,
+	sharedSecret sphinx.Hash256,
+	errorStruct *sphinx.AttributableErrorStructure) *SphinxErrorEncrypter {
+
+	encrypter := &SphinxErrorEncrypter{
+		EphemeralKey: ephemeralKey,
+		errorStruct:  errorStruct,
+	}
+
+	if errorStruct != nil {
+		// Set creation time rounded to nanosecond to avoid differences
+		// after serialization.
+		encrypter.CreatedAt = time.Now().Truncate(time.Nanosecond)
```

Comment:
nit: I think `time.Now()` already [returns the a value in nanoseconds](https://pkg.go.dev/time#Time) (which happens to be the max granularity too)

Also, given that we will return the [hold time in ms](https://github.com/lightning/bolts/pull/1044/files#diff-ad65f0beaac5cef88f5fd7a8b9ca36cbc5a790f4815b4b50b9ea794a55aaf012R1052) and some databases cannot have a max granularity for timestamps of microseconds (I am thinking about [postgres](https://www.postgresql.org/docs/12/datatype-datetime.html)) should we truncate to micro/miliseconds here?

Reply:
There was a reason why I had to do this. I think it was to strip off the monotonic clock reading.